java/src/main/java/ai/rapids/cudf/ParquetColumnWriterOptions.java

```diff
@@ -0,0 +1,539 @@
+/*
+ *
+ *  Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package ai.rapids.cudf;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.IntStream;
+
+/**
+ * Per column settings for writing Parquet files.
+ */
+public class ParquetColumnWriterOptions {
```

Comment:
I think that there is a lot that can be done to clean this code up. It looks like there is a lot of copy and paste and type hierarchy that I don't really understand.

For example, why does LeafBuilder need to be a ParquetWriterOptionsBuilder. It appears to be an internal implementation so why force ourselves to have the same APIs as the nested types and then end up using them in an inconsistent way with the nested types.  Why even have a builder at all, you should be able to skip it and just build the Options directly.

After that NestedBuilder can go away and just be a part of Builder, because all of the Builders are for Nested types (which is the public API). And in fact the core implementation of all the ListBuilder and the StructBuilder are almost identical, except the ListBuilder needs to check that there is only ever one child added, and we can either do that as a part of `build` which would end up being a lot less code, or we could add in a single method in Builder that would do the length check for the list but be a noop for the struct builder.

The also why is ParquetWriterOptions.Builder not also an instance of ParquetColumnWriterOptions.StructBuilder? It would make it so there is even less copy and paste because arguably the base Table being written is really just a Struct.

Reply:
Thanks for reviewing. I have updated the PR with more cleanup. I don't know why I didn't do it this way to start with, but this is looking good I think. I have gotten rid of the `ParquetWriterOptionsBuilder` interface and the `ParquetColumnWriterOptions` class. Everything is now a `ParquetWriterOptions`. PTAL I may missed some cleanup.

Reply:
It looks better, but I think it swung too far in the other direction.  I can now add Metadata, compression type, and statistics frequency to a ListBuilder and a StructBuilder.  I would prefer to see something like the following. I am fine if we have some of them be inner classes of others, but I think the hierarchy should look more like...

```
public class ParquetColumnOptions {
  private boolean isTimestampTypeInt96;
  private int precision;
  private boolean isNullable;
  private String columnName;
  private List<ParquetColumnOptions> childOptions;

  ParquetColumnOptions(...)
  ...

  public static class NestedBuilder<T extends NestedBuilder> {
    ...
  }
}

public ParquetListColumnOptions extends ParquetColumnOptions {
  ...
  public static class Builder extends ParquetColumnOptions.NestedBuilder<ParquetColumnOptions.NestedBuidler> {
    ...
  }
}

public ParquetStructColumnOptions extends ParquetColumnOptions {
  ...
  public static class Builder extends ParquetColumnOptions.NestedBuilder<ParquetColumnOptions.NestedBuidler> {
    ...
  }
}

public class ParquetWriterOptions extends ParquetStructColumnOptions {
  private CompressionType compressionType;
  private Map<String, String> metadata;
  private StatisticsFrequency statsGranularity;
  ...

   public static class Builder extends ParquetStructColumnOptions.Builder {
    ...
   withCompressionType();
   withMetadata();
   withStatsFrequency();
  }
}
```

Reply:
oh shoot! I meant to clean that up but missed it. I have not followed exactly what you have listed here but what I have done is renamed the old `Builder` to `NestedBuilder` and took out the `metadata`, `statisticsFrequency` and `compressionType` from it and created a new `Builder` that extends `NestedBuilder` along with the `metadata` etc in it. `StructBuilder` and `ListBuilder` extend `NestedBuilder`. PTAL when you get a chance

Reply:
I still think that  ParquetWriterOptions should be a ParquetStructColumnWriterOptions not the other way around.